### PR TITLE
update log recoverer to update the block time periodically

### PR DESCRIPTION
**Pull request by RyanRHall**
https://smartcontract-it.atlassian.net/browse/AUTO-4609

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go [hunk 1]
```diff
@@ -112,34 +114,40 @@ func (r *logRecoverer) Start(pctx context.Context) error {
...
 	blockTimeResolver := newBlockTimeResolver(r.poller)
-	blockTime, err := blockTimeResolver.BlockTime(ctx, defaultSampleSize)
-	if err != nil {
-		// TODO: TBD exit or just log a warning
-		// return fmt.Errorf("failed to compute block time: %w", err)
-		r.lggr.Warnw("failed to compute block time", "err", err)
-	}
-	if blockTime > 0 {
-		r.blockTime.Store(int64(blockTime))
+	updateBlockTime := func(ctx context.Context) {
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:117]: Could this function live on the logRecoverer, i.e. could it have a `(r *logRecoverer)` receiver, and take the blockTimeResolver as a param?

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go [hunk 2]
```diff
@@ -112,34 +114,40 @@ func (r *logRecoverer) Start(pctx context.Context) error {
...
-				case <-ticker.C:
+				case <-recoverTicker.C:
 					if err := r.recover(ctx); err != nil {
 						r.lggr.Warnw("failed to recover logs", "err", err)
 					}
 				case <-gcTicker.C:
 					r.clean(ctx)
 					gcTicker.Reset(utils.WithJitter(GCInterval))
+				case <-blockTimeUpdateTicker.C:
+					updateBlockTime(ctx)
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:150]: We might need to spawn a goroutine for the block time update as it could block (makes DB calls)
Same for `clean()` so we can also do it in a separate PR

**Review by amirylm — DISMISSED** [review]
LGTM

we can have another PR for the comment I wrote

**Review by infiloop2 — COMMENTED** [review]
can we also remove the todo from the file

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go [hunk 3]
```diff
@@ -112,34 +114,40 @@ func (r *logRecoverer) Start(pctx context.Context) error {
...
-	if blockTime > 0 {
-		r.blockTime.Store(int64(blockTime))
+	updateBlockTime := func(ctx context.Context) {
+		blockTime, err := blockTimeResolver.BlockTime(ctx, defaultSampleSize)
+		if err != nil {
+			r.lggr.Warnw("failed to compute block time", "err", err)
+			return
+		}
+		if blockTime > 0 {
+			r.blockTime.Store(int64(blockTime))
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:124]: can we add a debug log with the new block time here

Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:124]: and also maybe a warning if it changed by a lot?

Comment (RyanRHall) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:124]: Did I already remove the TODO? I don't see one.

Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:124]: ah yes, removed from line 117, my bad

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go [hunk 4]
```diff
@@ -650,3 +650,21 @@ func (r *logRecoverer) sortPending(latestBlock uint64) {
...
+func (r *logRecoverer) updateBlockTime(ctx context.Context) {
+	blockTime, err := r.blockTimeResolver.BlockTime(ctx, defaultSampleSize)
+	if err != nil {
+		r.lggr.Warnw("failed to compute block time", "err", err)
+		return
+	}
+	if blockTime > 0 {
+		currentBlockTime := r.blockTime.Load()
+		newBlockTime := int64(blockTime)
+		if (int64(math.Abs(float64(currentBlockTime-newBlockTime))) * 100 / currentBlockTime) > 20 {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:663]: will need to check for currentBlockTime > 0

**Review by infiloop2 — APPROVED** [review]

**Review by ferglor — APPROVED** [review]
